src/backend/distributed/transaction/transaction_management.c

```diff
@@ -483,8 +495,12 @@ static void
 PopSubXact(SubTransactionId subId)
 {
 	MemoryContext old_context = MemoryContextSwitchTo(CurTransactionContext);
-	Assert(linitial_int(activeSubXacts) == subId);
-	activeSubXacts = list_delete_first(activeSubXacts);
+	SubXactState *state = linitial(activeSubXactStates);
+
+	Assert(state->subId == subId);
+	activeSetStmts = state->setLocalCmd;
```

Comment:
I'm not sure I follow this logic. Should we not keep separate SET LOCAL lists for each subxact level?

Reply:
Writing an explanation in code comments, since I anticipated this would be the confusing part (I somewhat deviously/lazily like to see what code reviewers are confused by before explaining things in comments to see what I need to focus on).